teal/game/player.tl

```diff
@@ -0,0 +1,120 @@
+require("shared.init")
+
+local DEGREES_TO_RADIANS <const> = 3.1415 / 180
+local OSCILLATE_AMPLITUDE <const> = 2
+
+-- An object that represents a single bubble
+local record Bubble
+    canvas: Canvas.SubCanvas
+    previousSpriteIndex: integer
+    spriteIndex: integer
+    x: number
+    y: number
+    x_vel: number
+    y_vel: number
+    -- internal incremental ticker
+    ticker: number
+    ticker_speed: number
+    -- must be integral: the pixels into which to draw the bubble
+    draw_x: integer
+    draw_y: integer
+end
+
+local BubbleMetatable: metatable<Bubble> = {
+    __index = Bubble,
+ }
+
+function Bubble.new(): Bubble
+    local instance: Bubble = {}
+    setmetatable(instance, BubbleMetatable)
+    instance.canvas = Canvas.newCanvas(true)
+    instance.previousSpriteIndex = -1
+    instance.spriteIndex = 0
+    instance.x = 30
+    instance.y = config.displayHeight - 16
+    instance.x_vel = 0.2
+    instance.y_vel = 0.2
+    instance.ticker_speed = 1.1
+    instance.ticker = 0
+    instance.draw_x = 0
+    instance.draw_y = 0
+    return instance
+
+end
+
+-- keyStates: input key pressed or not
+-- dt: miliseconds
+function Bubble:step(dt: number, keyStates: {string:boolean})
+    local prevX = self.x
+    local prevY = self.y
+
+    if (keyStates["w"] and not keyStates["s"]) then
+        self.y = self.y - self.y_vel * dt
+    end
+    if (keyStates["s"] and not keyStates["w"]) then
+        self.y = self.y + self.y_vel * dt
+    end
+    if (keyStates["a"] and not keyStates["d"]) then
+        self.spriteIndex = 0
+
+        self.x = self.x - self.x_vel * dt
+    end
+    if (keyStates["d"] and not keyStates["a"]) then
+        self.spriteIndex = 1
+        self.x = self.x + self.x_vel * dt
+    end
+    -- increment the internal ticker for animations
+    self.ticker = math.fmod((self.ticker + dt) * self.ticker_speed, 360)
+
+    -- have to keep these seperate to avoid rounding errors
+    self.draw_x = math.floor(self.x + math.sin(self.ticker * DEGREES_TO_RADIANS) * OSCILLATE_AMPLITUDE)
+    self.draw_y = math.floor(self.y)
+end
+
+function Bubble:draw()
+    if (self.spriteIndex ~= self.previousSpriteIndex) then
+        self.canvas.clearCanvas()
+        self.canvas.drawImage("bubble.png", 0, 0, 64, 64, 0, 0, 64, 64)
+        self.previousSpriteIndex = self.spriteIndex
+    end
+    self.canvas.draw(self.draw_x, self.draw_y)
+end
+
+-- An object that represents everything that a player controls, along with it's
+-- canvas
+local record Player
+    bubbles: {Bubble}
+end
+
+local PlayerMetatable: metatable<Player> = {
+    __index = Player,
+ }
+
+function Player.new(): Player
+    local instance: Player = {}
+    setmetatable(instance, PlayerMetatable)
+    instance.bubbles = {}
+    table.insert(instance.bubbles, Bubble.new())
+    table.insert(instance.bubbles, Bubble.new())
+    instance.bubbles[2].x = 80
```

Comment:
Maybe we could pass these to the bubble initialiser?
You can do something like
```lua
function Bubble.new(x?: number, y?: number)
    -- existing stuff
    instance.x = x or 30
    instance.y = y or config.displayHeight - 16
    
```

Reply:
Will do so when it's needed.